web/packages/shared/hooks/useRefClickOutside/useRefClickOutside.ts

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { MutableRefObject, useEffect, useRef, useCallback } from 'react';
+
+/**
+ * useRefClickOutside adds a `mousedown` event listener (and cleanup)
+ * that upon `mousedown` will:
+ *   - set the field `open` to false if the event happened outside the ref
+ *   - does nothing if the event happened inside the ref
+ *
+ * Returns a `ref` object to be used for the element that we want `mousedown`
+ * events to be ignored.
+ */
+export function useRefClickOutside<
+  T extends { contains(eventTarget: HTMLElement): boolean }
+>(options: { open: boolean; setOpen(b: boolean): void }): MutableRefObject<T> {
+  const ref = useRef<T>();
+
+  const handleClickOutside = useCallback((event: MouseEvent) => {
+    if (ref.current && !ref.current.contains(event.target as HTMLElement)) {
```

Comment:
@gzdunek Do you remember if there was a significant difference between doing `element.contains(event.target)` vs `event.composedPath().includes(element)`?

https://github.com/gravitational/teleport/blob/6d3dd3c47deded4c878322db6f921ce42f8033fe/web/packages/teleterm/src/ui/Search/SearchBar.tsx#L80-L84